Async commit

sentientwaffle opened the pull request:
Add placeholder async functions `prefetch`/`compact`/`checkpoint` to `StateMachine`. Call them from `Replica`.
The replica must be careful about performing certain operations while it is busy committing.

**Superseded by https://github.com/coilhq/tigerbeetle/pull/127**

sentientwaffle committed:
StateMachine: Move into Replica.

This is a preliminary change for making `Replica.commit_op` async, so that the callback can use `@fieldParentPtr` to get the `replica`.

sentientwaffle committed:
VSR/StateMachine: Async commit hooks (WIP)

The replica changes are incomplete — the VOPR currently fails when the commit is delayed.

sentientwaffle committed:
VSR/StateMachine: Async commit hooks

Fix issues arising from async commit changes, including:

- Fix `Message` leak (in `Replica.deinit()`).
- Don't send a `do_view_change` message while committing.
- Don't start a view as the new leader while committing.
- After committing (in `commit_done`) resume sending the `do_view_change` or starting the view as the new leader, as appropriate.
- Move pop-head-of-pipeline into `commit_op_prefetch_callback` to maintain the `commit_min + 1 == pipeline.head.op` invariant.

In addition:

- In the simulator, vary the latency of `commit_{prefetch,compact,checkpoint}`.
- Add required parameters to `StateMachine.prefetch`.

sentientwaffle committed:
VSR: Async commit hooks: Fix pipeline check

In `commit_done`, only verify the pipeline if we are a leader in normal mode, since otherwise the pipeline isn't going to be committed anyway, and may be out of date.

Also included:

- Simulator cleanup: Remove the `commit_{prefetch,compact,checkpoint}_min` fields; they aren't useful.
- Improve validation after/around commit.
- In `flush_loopback_queue`, document a few missing cases in which a message is sent to self.

sentientwaffle committed:
VSR: Check invariants

sentientwaffle commented:
For symmetry what do you think of renaming:

- `commit_ops` → `commit_journal` (To mirror `commit_pipeline`, because `commit_ops` isn't used when committing the pipeline).
- `commit_ops_read` → `commit_journal_next` (To mirror `commit_pipeline_next`).
- `commit_ops_commit` → `commit_journal_next_callback`
- `commit_op_hooks` → `commit_op_prefetch` (to match the already-existing `commit_op_prefetch_callback`).

My main goal is to make it obvious which parts of the commit path:
- are used when committing from the WAL, vs
- those that are used when committing from the pipeline, vs
- those that are shared between both.

sentientwaffle commented:
It occurs to me that I introduced a bug in `on_do_view_change`:

A committing replica will defer sending the `do_view_change`. If that committing replica is the new leader, it sends to `do_view_change` to itself. But by the time it arrives, more than `threshold - 1` replicas may have already sent their `do_view_change` messages, so `count > threshold`.

https://github.com/coilhq/tigerbeetle/blob/2715a3ef41fc340c78364c3ba38f291fea2efb44/src/vsr/replica.zig#L1050-L1066

If the simulator hits this, the view change won't complete, but will eventually time out (i.e. no obvious VOPR failure.

---

Brainstorming possible fixes (the last one is my favorite so far):

- Skip the `count == threshold` checks. Instead, use `replica.do_view_change_quorum` as a guard.
- Or, ignore `do_view_change` messages that arrive before our own. (This would slow down view change due to necessary retries).
- Or, in `on_do_view_change` prior to `reference_message_and_receive_quorum_exactly_once` add something like:
```zig
if (self.do_view_change_from_all_replicas[self.replica] == null and
    self.replica != message.replica and
    (number of messages in do_view_change_from_all_replicas) + 1 == threshold)
{
    assert(self.committing);
    assert(!self.do_view_change_quorum);
    ... log
    return;
}
```

---

**Edit: It is safe to send a `do_view_change` while committing, which neatly fixes the above issue.**

jorangreef commented on src/simulator.zig:
```diff
@@ -115,6 +115,11 @@ pub fn main() !void {
+        .state_machine_options = .{
+            .commit_prefetch_mean = 5 + random.uintLessThan(u64, 10),
+            .commit_compact_mean = 5 + random.uintLessThan(u64, 10),
+            .commit_checkpoint_mean = 5 + random.uintLessThan(u64, 10),
```
Let's drop the `commit_` prefixes here, since these tasks are in series with (and on the same level as) `commit` itself.

i.e. It's like this: `prefetch() -> commit() -> compact() -> checkpoint()`

From VSR's perspective, we see this all as a logical commit, but from the state machine's perspective it is like `prepare/commit` is being expanded to `prepare/prefetch/commit/compact/checkpoint`.

jorangreef commented on src/simulator.zig:
```diff
@@ -243,6 +254,7 @@ pub fn main() !void {
             switch (cluster.health[replica.replica]) {
                 .up => |*ticks| {
                     ticks.* -|= 1;
+                    replica.state_machine.tick();
```
Perhaps `replica.state_machine.tick()` should be called directly from within `replica.tick()` rather?

sentientwaffle replied:
Right now `tick` is only on the simulator's `StateMachine` implementation (necessary for the mocked "async" functions), but not TB's primary `StateMachine` implementation.

jorangreef replied:
I think we may also need a `tick()` on `StateMachine` to call `state_machine.forest.tick()`.

cc @kprotty to confirm.

sentientwaffle replied:
:+1: If that's the case, I'll move the `tick` call into `Replica.tick`.

kprotty replied:
Sounds good.

jorangreef commented on src/state_machine.zig:
```diff
@@ -37,6 +37,12 @@ pub const StateMachine = struct {
         lookup_transfers,
     };
 
+    pub const Config = struct {
```
Let's rename this to `Options`.

jorangreef commented on src/state_machine.zig:
```diff
@@ -130,6 +134,32 @@ pub const StateMachine = struct {
+        _ = op_number;
+        callback(self);
+    }
+
```
Let's move this to be defined after `commit()` since that follows the logical order in which they are called.

jorangreef commented on src/test/state_machine.zig:
```diff
@@ -32,13 +67,48 @@ pub const StateMachine = struct {
     pub fn commit(
         state_machine: *StateMachine,
         client: u128,
+        op_number: u64,
```
Wouldn't be averse to just `op` here, since we use it like that in Replica and LSM.

jorangreef replied:
It's so core to the protocol and the type indicates that it's a number.

We'll also be clarifying all these state machine function interfaces with doc comments so it will be clear what it is.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -244,7 +255,7 @@ pub fn Replica(
             storage: *Storage,
             message_bus: *MessageBus,
-            state_machine: *StateMachine,
+            state_machine_config: StateMachine.Config,
```
Let's do `options` here too, then we don't overload on `config`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -295,6 +306,9 @@ pub fn Replica(
             errdefer journal.deinit(allocator);
 
+            var state_machine = try StateMachine.init(allocator, state_machine_config);
+            errdefer state_machine_config.deinit();
```
```suggestion
            errdefer state_machine.deinit(allocator);
```

We want to use Zig's `Managed` pattern here so that the state machine doesn't hold a reference to the allocator and therefore can't allocate dynamically at runtime, only during `init()`. Same as what we do for `Journal.init()/deinit()` above. This enforces static allocation quite nicely!

sentientwaffle replied:
I will get to this after I rebase.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -406,6 +421,13 @@ pub fn Replica(
+            if (self.commit_prepare) |committing_message| {
+                assert(self.committing);
+                assert(self.commit_callback != null);
+                self.message_bus.unref(committing_message);
```
```suggestion
            if (self.commit_prepare) |message| {
                assert(self.committing);
                assert(self.commit_callback != null);
                self.message_bus.unref(message);
```

As we do also below, for example, when we iterate the DVCs with `for (self.do_view_change_from_all_replicas) |message| {`

jorangreef replied:
Let's add the else branch to assert the negative space, i.e. that `commit_callback == null`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -154,6 +154,14 @@ pub fn Replica(
+        /// * start a view as the new leader
+        ///
+        /// because `commit_min` will be modified when the commit finishes.
         committing: bool = false,
```
Could we drop this boolean since we now have `commit_callback` as an optional that could be used as a guard?

sentientwaffle replied:
Unfortunately not. Right now the commit path works like:

1. Either:
    - Commit from WAL:
      1. `commit_ops`: Check `committing` guard. Either proceed or return.
      1. `commit_ops`: Set `committing = true`.
      1. `commit_ops_read`: Call `journal.read_prepare`.
      1. → Call `commit_op_hooks`.
    - Commit from pipeline:
      1. `commit_pipeline`: Check `committing` guard. Either proceed or return.
      1. `commit_pipeline`: Set `committing = true`.
      1. `commit_pipeline_next`: Call `commit_op_hooks`.
2. `commit_op_hooks`: Set `commit_callback` and `commit_prepare`, start prefetch
3. → prefetch done
4. start compact
5. → compact done, set `commit_callback = null` and `commit_prepare = null`, call (former) `commit_callback`.
6. The callback implementations (there are 2: one for commit-from-WAL, and one for commit-from-pipeline) will either:
    - decide to stop committing: call `commit_ops_done` (which sets `committing = false`)
    - or, jump to step 3 to commit another op.

(where "→" indicates that the step follows asynchronously).

If you find this hard to follow — I do too :sweat_smile:. But I'm leaning towards holding off refactoring for a separate PR to avoid holding up the LSM work.

One thing I think could help in the mean time is renaming some of the functions (suggestion: https://github.com/coilhq/tigerbeetle/pull/122#issuecomment-1199970696).

jorangreef replied:
100% agreed on a separate PR. Thanks for the detailed explanation here.

jorangreef replied:
The new names look good.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1894,7 +1939,7 @@ pub fn Replica(
 
             // It is critical that a `do_view_change` message implies a `start_view_change_quorum`:
-            if (self.start_view_change_quorum) {
+            if (self.start_view_change_quorum and !self.committing) {
```
Do we need this compound condition? Can we not keep the condition as is?

sentientwaffle replied:
It is unsafe without the guard.

The commit could take a really long time — long enough that the `view_change_message_timeout` fires before it finishes — and it isn't safe to `send_do_view_change` while `committing`.

sentientwaffle replied:
Update: it is safe to `send_do_view_change` while `committing` after all.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2235,31 +2284,111 @@ pub fn Replica(
+            self.commit_prepare = prepare.ref();
+            self.commit_callback = callback;
+            self.state_machine.prefetch(
+                self.commit_min + 1,
```
Let's pass `prepare.header.op` here rather since we've asserted it against `commit_min + 1` above already.

jorangreef replied:
This binds the `op` more closely to the prepare being committed.

jorangreef replied:
e.g. Similar to how we pass `prepare.header.operation` or `prepare.buffer`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2235,31 +2284,111 @@ pub fn Replica(
+
+            if (self.status == .normal and self.leader()) {
+                if (self.pipeline.pop()) |prepare| {
+                    assert(self.commit_min == self.commit_max);
```
Do we expect this outside of whether the pipeline has anything to pop?

jorangreef replied:
Do we not expect that the `pipeline.pop()` should always succeed?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2235,31 +2284,111 @@ pub fn Replica(
+
+            const commit_min_before = self.commit_min;
+            self.commit_op(self.commit_prepare.?);
+            assert(self.commit_min == commit_min_before + 1);
```
I think we could get rid of this local state variable, to shrink the scope, and also get a little bit stronger guarantees here, if we assert the relation between `commit_min` and `commit_prepare.header.op` before and after calling `commit_op`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2235,31 +2284,111 @@ pub fn Replica(
+                            // Write the next message in the queue.
+                            // A cluster-of-one writes prepares sequentially to avoid gaps in the
+                            // WAL caused by reordered writes.
+                            self.write_prepare(next.message, .append);
```
I guess there are no reentrancy issues here? i.e. That could feed back into another commit flow kicking off?

sentientwaffle replied:
I think the `committing` flag protects it, but lmk if I'm missing something.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2235,31 +2284,111 @@ pub fn Replica(
+                }
+            }
+
+            self.state_machine.compact(self.commit_min, commit_op_compact_callback);
```
I think it would be good to use `prepare.header.op` here consistently, since `commit_min` changes during the async hooks lifecycle. We should always be passing the same `op` number to these hooks, before and after committing through the state machine.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2335,49 +2464,86 @@ pub fn Replica(
+            assert(self.commit_min <= self.commit_max);
+            assert(self.commit_min <= self.op);
 
-                if (self.replica_count == 1) {
-                    if (self.pipeline.head_ptr()) |head| {
-                        // Write the next message in the queue.
-                        // A cluster-of-one writes prepares sequentially to avoid gaps in the WAL.
-                        self.write_prepare(head.message, .append);
-                        // The loop will wrap around and exit when `!ok_quorum_received`.
-                    }
```
Could you explain why we drop this now?

jorangreef replied:
Did this move elsewhere?

sentientwaffle replied:
This moved to `commit_op_prefetch_callback`. The `pipeline.pop()` part moved to enforce the `commit_min` vs `pipeline.head` invariant, and this moved along with it.

The `append` moved along with it — the alternative is to put it in `write_prepare_callback` which makes more logical sense, but since the pipeline head hasn't been popped yet at that point it might need to skip ahead in the pipeline (potentially multiple entries, which is tricky to trackk).

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2335,49 +2464,86 @@ pub fn Replica(
-                        self.write_prepare(head.message, .append);
-                        // The loop will wrap around and exit when `!ok_quorum_received`.
-                    }
+            if (self.status == .normal and self.leader()) {
```
Could there be a bug in this compound condition where a commit chain (`prefetch/commit/compact/checkpoint`) doesn't run through all the steps?

sentientwaffle replied:
Can you elaborate? This callback is passed to `commit_op_hooks`, and will be called when `compact` finishes (since we don't currently call `checkpoint` from `Replica`).

jorangreef replied:
I was wondering if, during a `commit_op_hooks` chain, it's possible for the status of the replica to change, so that it perhaps calls `prefetch()` and perhaps `commit()` etc. but not `compact()`. Or something along those lines. The invariant to preserve is that all hooks in the chain are all-or-nothing.

sentientwaffle replied:
Gotcha. `commit_pipeline_callback` or `commit_journal_callback` is called after compact finishes for each commit. 
So by the time it reaches this branch, `commit_pipeline_callback` has just finished a `prefetch→commit→compact` chain.

Afaik there is no way to only perform part of the chain (except for crashing mid-way through, of course).

jorangreef reviewed (commented):
Nice work @sentientwaffle!

sentientwaffle committed:
VSR,StateMachine: Async commit fixes (code review)

- Rename
- Cleanup
- Add `tick()` to `StateMachine`'s API. The simulator's `StateMachine` already implemented it, but now that it is an "official" function, it is invoked by `Replica.tick()`.
- In `commit_op_prefetch_callback`, a normal-mode leader can always pop the pipeline head. Since they are normal mode, everything prior to the pipeline was already committed.

sentientwaffle committed:
VSR: Fix DVC race with async commit

A committing replica will defer sending the `do_view_change`. If that committing replica is the new leader, it sends to `do_view_change` to itself. But by the time it arrives, more than `threshold - 11 replicas may have already sent their `do_view_change` messages, so `count > threshold`.

See https://github.com/coilhq/tigerbeetle/pull/122#issuecomment-1200343153 for more detail.

sentientwaffle committed:
VSR: Allow do_view_change while committing

This is safe (and much simpler!). Even if the new DVC fails to arrive, the old one has the same headers. If there were enough acks for it to be committed, the new leader will not be able to discard the header — it will survive the view change.

sentientwaffle committed:
VSR: Rename commit functions for symmetry

Rename `Replica`'s commit functions to emphasize the symmetry between the journal and pipeline commit code paths.

sentientwaffle committed:
VSR: Allow view start as leader while committing

This is safer, simpler, and faster than using a guard.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -306,8 +306,8 @@ pub fn Replica(
-            var state_machine = try StateMachine.init(allocator, state_machine_config);
-            errdefer state_machine_config.deinit();
+            var state_machine = try StateMachine.init(allocator, state_machine_options);
+            errdefer state_machine.deinit();
```
Just need to pass `allocator` in to `state_machine.deinit()`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2015,9 +1971,35 @@ pub fn Replica(
+
+                if (message.header.command == .do_view_change) {
+                    // A replica may resend a `do_view_change` with a different commit if it was
+                    // committing originally. Keep the one with the highest commit.
```
Is this necessary for correctness?

jorangreef replied:
Just to understand better. It's nice to have the `if/else` branches here explicitly showing that this can happen.

sentientwaffle replied:
It isn't necessary for correctness — I'll clarify that in the code. It's more that if we have a newer DVC anyways, we may as well use it.

Alternatively, we could just ignore the second message regardless, which would be a little less code.

jorangreef replied:
Let's keep the code. Great to specify the branches like that. And add the comment.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2933,7 +2897,7 @@ pub fn Replica(
         fn flush_loopback_queue(self: *Self) void {
-            // There are five cases where a replica will send a message to itself:
+            // There are four cases where a replica will send a message to itself:
             // However, of these five cases, all but one call send_message_to_replica().
```
`five` should also be `four` here.

jorangreef replied:
Do we need to update our static allocation calculation for messages in the message pool after these changes?

sentientwaffle replied:
There is an update for the `replica.commit_prepare`. But the loopback queue holds at most one message, so it should be fine.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -389,6 +396,7 @@ pub fn Replica(
         pub fn deinit(self: *Self, allocator: Allocator) void {
             self.journal.deinit(allocator);
             self.clock.deinit(allocator);
+            self.state_machine.deinit();
```
```suggestion
            self.state_machine.deinit(allocator);
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1069,6 +1088,24 @@ pub fn Replica(
+            // 1. Start committing op=N.
+            // 2. Send `do_view_change` to self.
+            // 3. Finish committing op=N.
+            // 4. Remaining `do_view_change` arrives, quorum complete.
```
```suggestion
            // 4. Remaining `do_view_change` messages arrive, quorum complete.
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -1069,6 +1088,24 @@ pub fn Replica(
+            // In this scenario, the our own `do_view_change`'s commit is `N-1`, but `commit_min=N`.
+            // Don't let the commit backtrack.
+            if (k.? < self.commit_min) {
+                assert(k.? == self.do_view_change_from_all_replicas[self.replica].?.header.commit);
```
Could it happen that our `commit_min` has increased, but another replica has increased yet further (but without sending another DVC to us)?

sentientwaffle replied:
Yes. But isn't that equivalent to a view-change in which the previous view's leader (with the highest commit of any replica) does not participate in the new view?

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -2742,6 +2924,8 @@ pub fn Replica(
             //    asynchronous prepare_ok to itself.
             // 3. In on_start_view_change(), after receiving a quorum of start_view_change
             //    messages, the new leader sends a synchronous do_view_change to itself.
+            // 4. In start_view_as_the_new_leader(), the new leader sends itself a prepare_ok
```
Great to have this comment here.

sentientwaffle committed:
VSR: Clarify documentation, code review

jorangreef commented:
LATM!

sentientwaffle commented:
Superseded by https://github.com/coilhq/tigerbeetle/pull/127.
